File: apps/studio/src/face/expressions.ts
import type { ExpressionId } from '@xiaomu/contracts';

/**
 * Parameters that drive the SVG2D face for a given expression.
 *
 * Coordinate system: viewBox 0 0 320 200.
 * Eye centers: L(105,85) R(215,85). Mouth center: (160,148).
 *
 * eyeRy is the "target" open ry; blink multiplies this to 0.
 * mouthCurve > 0 = smile (control point pulled down in SVG y).
 * mouthCurve < 0 = frown.
 * mouthOpen > 0 adds a filled jaw arc below the lip line.
 */
export interface EyeParams {
  rx: number;           // horizontal radius
  ry: number;           // vertical radius (open state)
  squintTop: number;    // 0=normal arc, 1=completely flat top (sleepy)
}

export interface MouthParams {
  width: number;        // half-width of mouth arc
  curve: number;        // control-point y offset: >0=smile, <0=frown
  open: number;         // jaw opening height (0=closed lips)
  round: boolean;       // O-shape for surprised/singing
}

export interface ExpressionDef {
  id: ExpressionId;
  label: string;
  color: string;         // glow + accent hex
  leftEye: EyeParams;
  rightEye: EyeParams;
  mouth: MouthParams;
  headTilt: number;      // degrees (positive = right lean)
  glowStrength: number;  // 1–3 multiplier on the drop-shadow
  idleBias: number;      // 0–1: how much idle modulation applies (0 = frozen intense)
  ssmlStyleHint: string;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

const eye = (rx: number, ry: number, squintTop = 0): EyeParams => ({ rx, ry, squintTop });
const mouth = (width: number, curve: number, open = 0, round = false): MouthParams => ({ width, curve, open, round });

// ── 16 expressions ───────────────────────────────────────────────────────────

export const EXPRESSIONS: Record<ExpressionId, ExpressionDef> = {
  happy: {
    id: 'happy',
    label: '开心',
    color: '#f59e0b',
    leftEye:  eye(22, 7),          // squinted from joy
    rightEye: eye(22, 7),
    mouth: mouth(34, 22, 2),       // wide smile, slightly open
    headTilt: 2,
    glowStrength: 2.5,
    idleBias: 0.6,
    ssmlStyleHint: 'cheerful',
  },

  excited: {
    id: 'excited',
    label: '兴奋',
    color: '#f97316',
    leftEye:  eye(19, 19),         // wide circular eyes
    rightEye: eye(19, 19),
    mouth: mouth(36, 22, 10),      // big open smile
    headTilt: -3,
    glowStrength: 3,
    idleBias: 0.3,
    ssmlStyleHint: 'excited',
  },

  calm: {
    id: 'calm',
    label: '平静',
    color: '#38bdf8',
    leftEye:  eye(20, 13),
    rightEye: eye(20, 13),
    mouth: mouth(28, 8, 0),        // gentle smile
    headTilt: 0,
    glowStrength: 1.2,
    idleBias: 1,
    ssmlStyleHint: 'calm',
  },

  gentle: {
    id: 'gentle',
    label: '温柔',
    color: '#6ee7b7',
    leftEye:  eye(18, 11),
    rightEye: eye(18, 11),
    mouth: mouth(24, 10, 0),       // soft small smile
    headTilt: 1,
    glowStrength: 1,
    idleBias: 1,
    ssmlStyleHint: 'gentle',
  },

  listening: {
    id: 'listening',
    label: '倾听',
    color: '#818cf8',
    leftEye:  eye(20, 15),
    rightEye: eye(20, 15),
    mouth: mouth(22, 4, 0),        // attentive neutral
    headTilt: 3,                   // slight lean-in tilt
    glowStrength: 1.5,
    idleBias: 0.8,
    ssmlStyleHint: 'gentle',
  },

  curious: {
    id: 'curious',
    label: '好奇',
    color: '#a3e635',
    leftEye:  eye(20, 18),         // left eye slightly bigger
    rightEye: eye(18, 14),         // asymmetric
    mouth: mouth(24, 7, 0),
    headTilt: -4,                  // quizzical tilt
    glowStrength: 1.8,
    idleBias: 0.7,
    ssmlStyleHint: 'curious',
  },

  thinking: {
    id: 'thinking',
    label: '思考',
    color: '#94a3b8',
    leftEye:  eye(19, 11),
    rightEye: eye(19, 11),
    mouth: mouth(20, 1, 0),        // pursed, nearly flat
    headTilt: -5,                  // looking away
    glowStrength: 1,
    idleBias: 0.5,
    ssmlStyleHint: 'calm',
  },

  sad: {
    id: 'sad',
    label: '难过',
    color: '#6366f1',
    leftEye:  eye(20, 9, 0.3),     // drooped tops
    rightEye: eye(20, 9, 0.3),
    mouth: mouth(28, -14, 0),      // frown
    headTilt: 1,
    glowStrength: 1,
    idleBias: 0.9,
    ssmlStyleHint: 'sad',
  },

  anxious: {
    id: 'anxious',
    label: '紧张',
    color: '#f43f5e',
    leftEye:  eye(17, 16),         // wide, slightly narrow
    rightEye: eye(17, 16),
    mouth: mouth(22, -5, 0),       // slight frown, tense
    headTilt: -2,
    glowStrength: 2,
    idleBias: 0.4,
    ssmlStyleHint: 'calm',
  },

  sleepy: {
    id: 'sleepy',
    label: '困倦',
    color: '#c084fc',
    leftEye:  eye(22, 6, 0.65),    // half-closed with flat top
    rightEye: eye(22, 6, 0.65),
    mouth: mouth(26, 3, 0),        // barely-there smile
    headTilt: 2,
    glowStrength: 0.8,
    idleBias: 1,
    ssmlStyleHint: 'calm',
  },

  surprised: {
    id: 'surprised',
    label: '惊讶',
    color: '#fbbf24',
    leftEye:  eye(18, 20),         // maximum vertical
    rightEye: eye(18, 20),
    mouth: mouth(14, 0, 14, true), // O-shape
    headTilt: 0,
    glowStrength: 3,
    idleBias: 0.2,
    ssmlStyleHint: 'excited',
  },

  celebrating: {
    id: 'celebrating',
    label: '庆祝',
    color: '#ec4899',
    leftEye:  eye(18, 17),
    rightEye: eye(18, 17),
    mouth: mouth(34, 22, 8),       // big open smile, similar to excited
    headTilt: -4,
    glowStrength: 3,
    idleBias: 0.2,
    ssmlStyleHint: 'excited',
  },

  proud: {
    id: 'proud',
    label: '自豪',
    color: '#a78bfa',
    leftEye:  eye(22, 12),         // wide, confident
    rightEye: eye(22, 12),
    mouth: mouth(30, 11, 0),       // composed smile
    headTilt: -2,                  // slight upward tilt
    glowStrength: 1.8,
    idleBias: 0.7,
    ssmlStyleHint: 'gentle',
  },

  confused: {
    id: 'confused',
    label: '困惑',
    color: '#fb923c',
    leftEye:  eye(20, 14),
    rightEye: eye(16, 11, 0.2),    // right eye squinted/raised
    mouth: mouth(20, -2, 0),       // crooked, slight downward
    headTilt: 5,
    glowStrength: 1.5,
    idleBias: 0.6,
    ssmlStyleHint: 'calm',
  },

  playful: {
    id: 'playful',
    label: '俏皮',
    color: '#34d399',
    leftEye:  eye(22, 8),          // squinted wink-like
    rightEye: eye(22, 14),         // one more open
    mouth: mouth(30, 17, 2),       // big grin
    headTilt: -3,
    glowStrength: 2,
    idleBias: 0.5,
    ssmlStyleHint: 'cheerful',
  },

  encouraging: {
    id: 'encouraging',
    label: '鼓励',
    color: '#60a5fa',
    leftEye:  eye(20, 13),
    rightEye: eye(20, 13),
    mouth: mouth(30, 15, 0),       // warm confident smile
    headTilt: 1,
    glowStrength: 1.8,
    idleBias: 0.9,
    ssmlStyleHint: 'cheerful',
  },
};

export const EXPRESSION_LIST = Object.values(EXPRESSIONS);

File: robot/face/expressions.py
"""The 16 expression poses — a faithful Python port of
apps/studio/src/face/expressions.ts.

These are the SAME poses the studio's SVG2DRenderer draws, in the SAME
320x200 design space (eye centers L(105,85) R(215,85), mouth (160,148)).
The robot's PythonLCD renderer (face/renderer.py) is a second implementation
of the project's pluggable FaceRenderer (CLAUDE.md §12), driven by this data.

Colors here match the studio defaults; at runtime a published StudioBundle's
config.face.expressionLibrary[id].colorHex overrides them (see face/bundle hook).
"""
from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Dict, List

# Design-space landmarks (viewBox 0 0 320 200) — keep in sync with the SVG renderer.
VIEW_W, VIEW_H = 320, 200
L_EYE = (105, 85)
R_EYE = (215, 85)
MOUTH = (160, 148)
PANEL_BG = (0x1A, 0x14, 0x20)  # #1a1420


@dataclass(frozen=True)
class Eye:
    rx: float          # horizontal radius
    ry: float          # vertical radius (open state)
    squint_top: float = 0.0  # 0=normal, 1=fully flat top (sleepy)


@dataclass(frozen=True)
class Mouth:
    width: float       # half-width of the mouth arc
    curve: float       # control-point y offset: >0 smile, <0 frown
    open: float = 0.0  # jaw opening height (0 = closed lips)
    round: bool = False  # O-shape for surprised / singing


@dataclass(frozen=True)
class Expression:
    id: str
    label: str
    color: str         # hex "#rrggbb"
    left_eye: Eye
    right_eye: Eye
    mouth: Mouth
    head_tilt: float   # degrees (positive = right lean)
    glow_strength: float  # 1-3 multiplier on the glow
    idle_bias: float   # 0-1: how much idle modulation applies
    ssml_style_hint: str


def _e(rx: float, ry: float, squint: float = 0.0) -> Eye:
    return Eye(rx, ry, squint)


def _m(width: float, curve: float, open_: float = 0.0, round_: bool = False) -> Mouth:
    return Mouth(width, curve, open_, round_)


EXPRESSIONS: Dict[str, Expression] = {
    "happy":       Expression("happy", "开心", "#f59e0b", _e(22, 7), _e(22, 7), _m(34, 22, 2), 2, 2.5, 0.6, "cheerful"),
    "excited":     Expression("excited", "兴奋", "#f97316", _e(19, 19), _e(19, 19), _m(36, 22, 10), -3, 3, 0.3, "excited"),
    "calm":        Expression("calm", "平静", "#38bdf8", _e(20, 13), _e(20, 13), _m(28, 8, 0), 0, 1.2, 1, "calm"),
    "gentle":      Expression("gentle", "温柔", "#6ee7b7", _e(18, 11), _e(18, 11), _m(24, 10, 0), 1, 1, 1, "gentle"),
    "listening":   Expression("listening", "倾听", "#818cf8", _e(20, 15), _e(20, 15), _m(22, 4, 0), 3, 1.5, 0.8, "gentle"),
    "curious":     Expression("curious", "好奇", "#a3e635", _e(20, 18), _e(18, 14), _m(24, 7, 0), -4, 1.8, 0.7, "curious"),
    "thinking":    Expression("thinking", "思考", "#94a3b8", _e(19, 11), _e(19, 11), _m(20, 1, 0), -5, 1, 0.5, "calm"),
    "sad":         Expression("sad", "难过", "#6366f1", _e(20, 9, 0.3), _e(20, 9, 0.3), _m(28, -14, 0), 1, 1, 0.9, "sad"),
    "anxious":     Expression("anxious", "紧张", "#f43f5e", _e(17, 16), _e(17, 16), _m(22, -5, 0), -2, 2, 0.4, "calm"),
    "sleepy":      Expression("sleepy", "困倦", "#c084fc", _e(22, 6, 0.65), _e(22, 6, 0.65), _m(26, 3, 0), 2, 0.8, 1, "calm"),
    "surprised":   Expression("surprised", "惊讶", "#fbbf24", _e(18, 20), _e(18, 20), _m(14, 0, 14, True), 0, 3, 0.2, "excited"),
    "celebrating": Expression("celebrating", "庆祝", "#ec4899", _e(18, 17), _e(18, 17), _m(34, 22, 8), -4, 3, 0.2, "excited"),
    "proud":       Expression("proud", "自豪", "#a78bfa", _e(22, 12), _e(22, 12), _m(30, 11, 0), -2, 1.8, 0.7, "gentle"),
    "confused":    Expression("confused", "困惑", "#fb923c", _e(20, 14), _e(16, 11, 0.2), _m(20, -2, 0), 5, 1.5, 0.6, "calm"),
    "playful":     Expression("playful", "俏皮", "#34d399", _e(22, 8), _e(22, 14), _m(30, 17, 2), -3, 2, 0.5, "cheerful"),
    "encouraging": Expression("encouraging", "鼓励", "#60a5fa", _e(20, 13), _e(20, 13), _m(30, 15, 0), 1, 1.8, 0.9, "cheerful"),
}

EXPRESSION_IDS: List[str] = list(EXPRESSIONS.keys())


def with_color(expr: Expression, color_hex: str) -> Expression:
    """Return a copy of expr with its color overridden (from a StudioBundle)."""
    return replace(expr, color=color_hex)
